File: src/shared/nonce.ts
/** tee-ai-nonce/v1: attestation nonce derived from the committed run. */

import { bytesToHex, concatBytes, hex32, sha256Bytes, utf8 } from "./canonical.js";

export const NONCE_SCHEME = "tee-ai-nonce/v1";
const DOMAIN = utf8("tee-ai-nonce/v1");

export interface AttestationNonceInputs {
  resultsRoot: string;
  datasetHash: string;
  registryHash: string;
  modelCommitment: string;
  policyHash: string;
  publicKeyFingerprint: string;
}

export async function deriveAttestationNonce(inputs: AttestationNonceInputs): Promise<string> {
  return bytesToHex(
    await sha256Bytes(
      concatBytes(
        DOMAIN,
        hex32(inputs.resultsRoot, "resultsRoot"),
        hex32(inputs.datasetHash, "datasetHash"),
        hex32(inputs.registryHash, "registryHash"),
        hex32(inputs.modelCommitment, "modelCommitment"),
        hex32(inputs.policyHash, "policyHash"),
        hex32(inputs.publicKeyFingerprint, "publicKeyFingerprint")
      )
    )
  );
}

File: src/model/tee_runner/nonce.py
"""tee-ai-nonce/v1: attestation nonce derived from the committed run.

Node computes this before asking the TEE for a token; the runner only needs
it for test vectors and self-checks.
"""

from __future__ import annotations

from .canonical import sha256_bytes

SCHEME = "tee-ai-nonce/v1"
DOMAIN = b"tee-ai-nonce/v1"


def attestation_nonce(
    results_root: str,
    dataset_hash: str,
    registry_hash: str,
    model_commitment: str,
    policy_hash: str,
    public_key_fingerprint: str,
) -> str:
    material = DOMAIN + b"".join(
        bytes.fromhex(part)
        for part in (
            results_root,
            dataset_hash,
            registry_hash,
            model_commitment,
            policy_hash,
            public_key_fingerprint,
        )
    )
    return sha256_bytes(material).hex()
